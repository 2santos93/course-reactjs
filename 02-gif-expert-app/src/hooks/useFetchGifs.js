import { useEffect, useState } from "react";

import { getGifs } from "../Services/gifs";

export const useFetchGifs = (category) => {

    const [state, setState] = useState({gifs:[], loading:true});

    useEffect(() => {
        async function getGifsData(){
            
            const gifsData = await getGifs({q:category});
            
            setState(
                {
                    loading:false,
                    gifs:[
                        ...gifsData,
                        ...state.gifs
                    ]
                }
            )
        }

        getGifsData();
    }, [category]);

    return state;
}