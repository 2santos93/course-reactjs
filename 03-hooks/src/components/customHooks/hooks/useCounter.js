import {useState} from 'react';

export const useCounter = (initialCounter=1) => {
    const [counter, setCounter] = useState(initialCounter);

    const increment = (e) => {
        setCounter(counter+1)
    }

    return [counter, increment];
}
