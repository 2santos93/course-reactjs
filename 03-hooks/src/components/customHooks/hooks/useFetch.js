import {useState, useEffect} from 'react';

export const useFetch = (url) => {
    
    const [state, setState] = useState({loading: true, data: null});

    useEffect(() => {
        const getQuotes = async () => {
            const response = await fetch(url);
            const data = await response.json();
            setState({loading: false, data});
        }
        setState({...state, loading:true})
        getQuotes();

    }, [url]);

    return state;
}
