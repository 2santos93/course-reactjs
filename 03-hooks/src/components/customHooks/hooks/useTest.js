import { useEffect } from "react"

export const useTest = (load) => {

  useEffect(() => {
    console.log('me monte' +load);
    return () => {
      console.log('me desmonte'+ load);
    }
  }, [])

}
