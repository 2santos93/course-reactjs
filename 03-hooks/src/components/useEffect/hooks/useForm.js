import { useState } from "react"

export const useForm = (initialState={}) => {
    const [state, setState] = useState(initialState);
    
    // const onSubmitHandler = (e) => {
    //     e.preventDefault();
    //     console.log(e);
    // }

    const onChangeHandler = (e) => {
        const {name, value} = e.target;
        setState((prev) => ({
            ...prev,
            [name]: value
        }))
    }
    
    return [state, onChangeHandler];
}
