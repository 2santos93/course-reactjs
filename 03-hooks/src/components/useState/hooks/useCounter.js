import { useState } from "react";

const useCounter = (initialState = 10) => {

    const [state, setstate] = useState(initialState)

    const increment = (factor = 1) => {
        setstate(state + factor);
    }
    const decrement = (factor = 1) => {
        setstate(state - factor);
    }
    const reset = () => {
        setstate(initialState);
    }

    return {decrement, increment, reset, state};
}

export {useCounter};
