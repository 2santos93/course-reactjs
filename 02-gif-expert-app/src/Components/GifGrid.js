import PropTypes from 'prop-types';
import { useFetchGifs } from "../hooks/useFetchGifs";
import GifGridItem from "./GifGridItem";

const GifGrid = ({category}) => {

    const {gifs, loading} = useFetchGifs(category);
    
    return (
        <div className="animate__animated animate__fadeIn">
            <h1>{category}</h1>

            {loading && <h2 className="animate__animated animate__heartBeat">Loading</h2>}

            <div className="card-container">
            {
                gifs.map((gif) => <GifGridItem key={gif.id} {...gif}/>)
            }
            </div>
        </div>

    );
}

GifGrid.propTypes = {
    category: PropTypes.string.isRequired
}

export default GifGrid;