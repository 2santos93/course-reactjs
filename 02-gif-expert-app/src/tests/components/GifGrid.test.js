import {shallow} from 'enzyme';

import GifGrid from '../../Components/GifGrid';
import {useFetchGifs} from '../../hooks/useFetchGifs';
jest.mock('../../hooks/useFetchGifs');

describe('test for GifGrid component', () => {
    const category = 'Dragon Ball';
    
    test('should render ok', () => {
        useFetchGifs.mockReturnValue({loading:true, gifs: []});
        const wrapper = shallow(<GifGrid category={category} />);
        const h2 = wrapper.find('h2');
        const GifGridItem = wrapper.find('GifGridItem');
        
        expect(wrapper).toMatchSnapshot();
        expect(h2.text()).toBe('Loading');
        expect(GifGridItem.exists()).toBe(false);
    });

    test('should show loading and not render the GifGridItem', () => {
        const gifs = [{id:'123', url:'http://localhost', title:'Hi im goku'}];
        
        useFetchGifs.mockReturnValue({loading:false, gifs});
        const wrapper = shallow(<GifGrid category={category} />);
        const GifGridItem = wrapper.find('GifGridItem');
        const h2 = wrapper.find('h2');

        expect(h2.exists()).toBe(false);
        expect(GifGridItem.exists()).toBe(true);
        expect(GifGridItem.length).toBe(gifs.length);

    });
    
})
