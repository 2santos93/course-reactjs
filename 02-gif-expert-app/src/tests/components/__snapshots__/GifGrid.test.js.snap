// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test for GifGrid component should render ok 1`] = `
<div
  className="animate__animated animate__fadeIn"
>
  <h1>
    Dragon Ball
  </h1>
  <h2
    className="animate__animated animate__heartBeat"
  >
    Loading
  </h2>
  <div
    className="card-container"
  />
</div>
`;
